### ISSUE TITLE
Make etcd self-hosted

### ISSUE THREAD

**Issue by johananl** [kinvolk/lokomotive#461]
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/cluster-lifecycle/self-hosted-kubernetes.md

Currently we are almost 100% self hosted ("level 0" in the design doc above). AFAICT the only remaining control plane element which isn't self hosted is **etcd**. Making etcd self hosted would give us the following benefits:

- Ability to upgrade etcd without replacing nodes
- A 100% self-hosted k8s distro (nice for marketing)
- A consistent design

I know there are other distros out there which self-host etcd (kops does IIRC), so I'm sure it's feasible. My hope is that it's not as tricky as self-hosting the kubelet, i.e. maybe we don't need a "fallback" etcd process which activates when the pod-based etcd is dead.

Making etcd self hosted is one possible answer to https://github.com/kinvolk/lokomotive/issues/161.

**Comment by rata** [comment]
No, kops considered doing it but didn't. It uses etcd-manager (https://github.com/kopeio/etcd-manager) that is not self-hosted (they decided to avoid the circular dependency that others, like etcd-operator, etc. would create).

IMHO, the reasons you listed are not good enough to go the self-hosted route for etcd. IMHO, we should:
 * See what other are doing and learn from their experience (like look at kops, I think others are running it as a container but not a pod, just a docker container, for example).
 * Understand the disaster recovery scenarios for different solutions (specially self-hosted)
 * From a technical POV, understand the pros/cons of each one and decide what is best

In my opinion, this can be done in a design doc with a clear proposal and is **specially important** to see the alternatives considered to self-hosting and why we decide to discard them (if we decide to discard them).

**Comment by invidian** [comment]
I was also considering having etcd self-hosted, but then dropped the idea, as disaster recovery might be more difficult than having it on the node level. We could perhaps run it as a static pod though.

**Comment by johananl** [comment]
If the cons outweigh the pros then maybe we should choose an alternative approach. For some reason I thought this was trivial as I'm sure I've seen etcd pods in the past (if not in kops then somewhere else). Anyway, I haven't evaluated this at all so please treat this issue as a "placeholder". At the very least we should be able to say "we aren't self-hosting etcd because [reasons]. Here is how you upgrade it."
